### PR TITLE
feat : script to automate the process of backup between mongodb and a…

### PR DESCRIPTION
**Pull request by hamzasysh**
…ws s3 bucket

### CODE REVIEW

File: script.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        try:
+            subprocess.run(command, check=True)
+            print("Backup completed successfully.")
+        except subprocess.CalledProcessError as e:
+            print(f"Backup failed: {e}")
+
+
+if __name__ == "__main__":
+
+    database = sys.argv[1]
```
Comment (zubair-0) [inline script.py]: use named arguments
or even better
make an .env file and keep the env vars there. then read the env vars from that file

File: script.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    obj_paths=sys.argv[5].split(":")
+    s3objpaths=sys.argv[6].split(",")
+    uri=sys.argv[7]
+    database2=sys.argv[8]
+
+
+    mongodump(database, collections,outpaths)
+    uploadtos3(outpaths,bucket,obj_paths)
+    download_from_s3(bucket, obj_paths, s3objpaths)
+    mongorestore(uri,database2,collections,s3objpaths)
```
Comment (zubair-0) [inline script.py]: in some places, there are spaces after commas and somewhere not
use good practices
start using a linter in your code like pyblack

File: script.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+def uploadtos3(outpaths,bucket,obj_paths):
+    try:
+        s3 = boto3.client('s3')
+    except Exception as e:
+        print(f"s3 bucket client failed: {e}")
+
+    for i in range(len(outpaths)):
+        try:
+            with open(outpaths[i], 'rb') as f:
```
Comment (zubair-0) [inline script.py]: you don't need to open the file to upload it
use the `upload_file` method by following the [documentation](https://boto3.amazonaws.com/v1/documentation/api/latest/guide/s3-uploading-files.html)

File: script.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
+import subprocess
+import sys
+import boto3
+
+def mongodump(database,collections,outpaths):
+     for i in range(len(collections)):
+          command = ["mongodump", "--uri", database, "--collection", collections[i],"--out",outpaths[i]]
```
Comment (zubair-0) [inline script.py]: if you are going to dump the db completely, no need to use `--collection`

File: script.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+def download_from_s3(bucket, obj_paths, s3objpaths):
+    # Initialize the S3 client
+    try:
+        s3 = boto3.client('s3')
+    except Exception as e:
+        print(f"s3 bucket client failed: {e}")
+
+    for i in range(len(obj_paths)):
+        try:
+            with open(s3objpaths[i], 'wb') as f:
```
Comment (zubair-0) [inline script.py]: you don't need to open the file to download it
use the `download_file` method by following the [documentation](https://boto3.amazonaws.com/v1/documentation/api/latest/guide/s3-example-download-file.html)

File: script.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+     for i in range(len(collections)):
+          command = ["mongodump", "--uri", database, "--collection", collections[i],"--out",outpaths[i]]
+          try:
+              subprocess.run(command, check=True)
+              print("Backup completed successfully.")
+          except subprocess.CalledProcessError as e:
+              print(f"Backup failed: {e}")
+
+def uploadtos3(outpaths,bucket,obj_paths):
+    try:
```
Comment (zubair-0) [inline script.py:20]: unnecessary try except block, see the docs

File: script.py [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+        try:
+            with open(outpaths[i], 'rb') as f:
+                s3.upload_fileobj(f, bucket, obj_paths[i])
+            print(f"File uploaded successfully to S3: s3://{bucket}/{obj_paths[i]}")
+        except Exception as e:
+            print(f"File upload to S3 failed: {e}")
+
+def download_from_s3(bucket, obj_paths, s3objpaths):
+    # Initialize the S3 client
+    try:
```
Comment (zubair-0) [inline script.py:33]: unnecessary try except block, see the docs

File: script.py [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+                s3.download_fileobj(bucket, obj_paths[i], f)
+            #s3.download_file(bucket, obj_paths[i],s3objpaths[i])
+            print(f"File downloaded successfully to: {s3objpaths[i]}")
+        except Exception as e:
+            print(f"File download from S3 failed: {e}")
+
+
+def mongorestore(uri,database2,collections,s3objpaths):
+    for i in range(len(collections)):
+        command = ["mongorestore", "--uri", uri, "--nsInclude", database2+"."+collections[i],s3objpaths[i]]
```
Comment (zubair-0) [inline script.py]: no need for `nsInclude` if you're going to restore the complete db

File: script.py [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+        print(command)
```
Comment (zubair-0) [inline script.py]: use logging library everywhere instead of print

**Review by zubair-0 — CHANGES_REQUESTED** [review]

File: script.py (hunk 1 above)
Comment (hamzasysh) [reply script.py]: to make the script dynamic , i didnt use .env file

File: script.py (hunk 3 above)
Comment (hamzasysh) [reply script.py]: both of them are identical. i just used this method. i can change if you want me to

File: script.py (hunk 4 above)
Comment (hamzasysh) [reply script.py]: no i am dumping it collection by collection

File: script.py (hunk 5 above)
Comment (hamzasysh) [reply script.py]: its actually a writable file object to write downloaded file to.

File: script.py (hunk 6 above)
Comment (hamzasysh) [reply script.py:20]: i did add try-except to boto3 client connection to see if connection is successful or not

File: script.py (hunk 7 above)
Comment (hamzasysh) [reply script.py:33]: i did add try-except to boto3 client connection to see if connection is successful or not

File: script.py (hunk 8 above)
Comment (hamzasysh) [reply script.py]: i am restoring it collection by collection.

File: script.py (hunk 9 above)
Comment (hamzasysh) [reply script.py]: i will remove these print statements. i added them fro debugging purposes

File: script.py (hunk 1 above)
Comment (hamzasysh) [reply script.py]: done

File: script.py (hunk 2 above)
Comment (hamzasysh) [reply script.py]: formatting done using pyblack

File: script.py (hunk 3 above)
Comment (hamzasysh) [reply script.py]: done

File: script.py (hunk 2 above)
Comment (hamzasysh) [reply script.py]: done

File: script.py (hunk 5 above)
Comment (hamzasysh) [reply script.py]: done

File: script.py (hunk 7 above)
Comment (hamzasysh) [reply script.py:33]: removed

File: script.py (hunk 8 above)
Comment (hamzasysh) [reply script.py]: i am restoring collection by collection

File: script.py (hunk 9 above)
Comment (hamzasysh) [reply script.py]: done

**Pull request merged**